### ISSUE TITLE
Loss? or Acc?

### ISSUE THREAD

**Issue by hundredeuk2** [BoocampMonster/korean-relation-extraction#4]
### Experiment status:
- Data : entity_marker_sentencepiece / Train_test_split(8:2)
- Model = EncoderModel (klue/roberta-large with biGRU 3 layers with Tanh)
- Tokenizer = add special tokens ['<E1>','</E1>','<E2>','</E2>']
- Epoch = 5

---

![Image](https://user-images.githubusercontent.com/73874591/203963990-361e0db5-e298-4373-a02e-74dc766fc85b.png)

1. Min Loss .pt
- Train/dev Acc & Loss : 0.8238 / 0.5211
- Submission Acc : 67.9630 / 71.2291

1. Max Acc .pt
- Train/dev Acc & Loss : 0.8309 / 0.5638
- Submission Acc : 68.1384 / 68.9010

**Comment by yhkee0404** [comment]
여기 ignore_index가 적용하기 딱인 옵션인 것 같아 수식도 있어: https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss

여기 설명도 있음: https://gaussian37.github.io/vision-segmentation-one_hot_label/

그 외에는 저 함수가 왜 좋은지 설명한 곳이 있었는지 당장은 tistory가 나가서 어떤 건지 못 찾겠다..

**Comment by hundredeuk2** [comment]
> 여기 ignore_index가 적용하기 딱인 옵션인 것 같아 수식도 있어: https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss
> 
> 여기 설명도 있음: https://gaussian37.github.io/vision-segmentation-one_hot_label/
> 
> 그 외에는 저 함수가 왜 좋은지 설명한 곳이 있었는지 당장은 tistory가 나가서 어떤 건지 못 찾겠다..

`ignore_index ([int](https://docs.python.org/3/library/functions.html#int), optional) – Specifies a target value that is ignored and does not contribute to the input gradient. When size_average is True, the loss is averaged over non-ignored targets. Note that ignore_index is only applicable when the target contains class indices.` 이부분 이군요 한번 참고해서 log 띄워볼께요 감사합니다!

**Comment by yhkee0404** [comment]
문제 푸는데 틀렸어. 근데 그게 수능 기출 문제야. 그럼 엄청 복습하려고 하겠지. 맞아도 또 엄청 복습하겠지.

근데 문제 푸는데 틀렸어. 알고 보니 사설 문제야. 그럼 대수롭지 않게 넘길 수 있잖아. 맞아도 어차피 수능도 아니라 별로고. 중요한 건 실전에서 미끄러지지 않는 거잖아. `Precision`이나 `Recall` 둘 중 하나만 집중하는 것은 이런 학습 방법인 것 같아.

검색하면 `F1`이나 `AURPC`나 종합적으로 골고루 보기 때문에 그냥 전체에서 정답 비율 구하는 단순 `Accuracy`보다 장점이 있다고 하지. 그런데 이 대회 `F1`은 `no relation`에 대해서는 골고루 보는 대회가 아님. 그걸 제외하고 골고루 보는 특이한 메트릭이다.

이렇게 일단 질러 보고 진짜면 초집중하고 가짜면 대충 넘기는 그런 게임 있을 것 같은데 예시를 잘 못 찾겠다.. 아무튼 성능 평가 지표를 `Precision`이나 `Recall`에 집중할 뿐 아니라 학습 시에도 수능 문제 예시처럼 대상 로스를 다르게 학습하는 게 전략이 될 수 있을 것 같아.

**Comment by yhkee0404** [comment]
저기 @gaussian37님 블로그는 한발 더 나아가서 `no relation`을 `1`개보다 많이 `20`개 `class`로 분류하고 무시하고 있다. `30`개 중에 `10`개만 살리고 있어. 이진 분류기 역할을 키워준 느낌이다..

**Comment by yhkee0404** [comment]
여기서 아이디어를 얻어서 좀더 설명할 수 있을 것 같아.

마술사가 다이아와 하트 각각 `10`개씩 총 `20`개만 있는 카드 중에서 나한테 뒤집힌 카드를 `1`장씩 주는데, 숫자가 짝수일 때 내가 다이아인지 하트인지 맞히면 나한테 `10`만원씩을 주겠대. 패스도 할 수 있어.

그러면 나는 뒤집힌 카드의 숫자가 홀수이면 돈 못 받는 헛수고하는 거 어떡하지 그럼 패스할까 하는 고민할 필요가 사실 없어. 무조건 다이아인지 하트인지만 `2`개 중에서만 부르면 되는 거야. 이때 숫자가 홀수이면 어쩔 수 없지만 짝수이면 무조건 이득이니까. 혹시나 홀수일까봐 주저하고 다이아인지 하트인지 말하지 않고 "혹시 홀수..? 그럼 의심 되니 패스" 같이 대답하는 사람은 바보인 거야.

카드를 뒤집어 보니 홀수야. 그러면 그 카드의 특징을 학습하는 게 좋을까? 홀수 다이아 카드가 파란색이었으니 앞으로 파란색은 다이아라고 외치면 좋을까? 아니야, 홀수의 파란색 분포는 중요하지 않고 짝수의 파란색 분포만 중요하기 때문에 학습하면 오히려 짝수 다이아 구분하는 데 방해 받을 위험까지 생겨.

카드를 뒤집어 보니 짝수야. 그러면 그 카드의 특징을 학습하는 게 좋지. 짝수 다이아 카드가 대부분 낡아 보인다는 특징을 찾아냈다면 앞으로 낡아 보이는 카드는 다이아라고 예측하는 것은 좋은 전략이야. 한편 만일 홀수 다이아 카드가 다 새 카드였다면 홀수 다이아 카드 문제는 엄청 틀리겠지. 그래도 홀수는 어차피 상관 없이 짝수를 맞히는 게 중요해.

위에서 `no relation` 또는 `20`개 `class`는 홀수 카드라서 그게 `20`가지 중 어떤 특징의 `no relation`인지는 관심이 없어. 나는 짝수 즉 `yes relation`의 특징을 학습한 것에 따라서만 분류하는 게 무조건 이득이야. 그래서 반대로 `yes relation` 또는 나머지 `10`개 `class`는 짝수 카드라서 그게 `10`가지 중 어떤 특징의 `yes relation`인지는 초미의 관심사지. 그걸 맞힐 때만 10만원씩 얻으니까.
